man/RNAplfold.1.in
.\" .ER
.TH "RNAPLFOLD" "l" "1.8" "Stephan Bernhart" "ViennaRNA"
.SH "NAME"
RNAplfold \- calculate locally stable secondary structure - pair probabilities 
.SH "SYNOPSIS"
\fBRNAplfold\fP [\-L span] [-W winsize] [\-T\ \fItemp\fP] [\-4] [\-noLP] [\-noGU]
[\-noCloseGU] [\-P\ \fIparamfile\fP] [\-nsp\ \fIpairs\fP]
[\-d[0|2] [\-c \fIcutoff\fP] [\-o] [\-u  \fInumber of unpaired\fP] [\-O]

.SH "DESCRIPTION"
.I RNAplfold
computes local pair probabilities for base pairs with a maximal span
of L. The probabilities are averaged over all windows of size L that
contain the base pair. For a sequence of length n and a window size of L the
algorithm uses only O(n+L*L) memory and O(n*L*L) CPU time. Thus it is
practical to "scan" very large genomes for short stable RNA structures.

Output consists of a dot plot in postscript file, where the averaged pair probabilities
can easily be parsed and visually inspected.

The -u option makes i possible to compute the probability
that a stretch of x consequtive nucleotides is unpaired, which is useful
for predicting possible binding sites. Again this probability is averaged over all windows containing the region.
.TP
.B WARNING! Output format changed!!
.PP
The output is a plain text matrix containing on each line a position i
followed by the probability that i is unpaired, [i-1..i] is unpaired [i-2..i]
is unpaired and so on to the probability that [i-x+1..i] is unpaired.  
 

.SH "OPTIONS"
.TP
.B \-W winsize
Average pair probabilities over windows of size \fIwinsize\fP. Default 70.
.PP
.TP
.B \-L span
Allow only pairs (i,j) with j\-i<=span. Default winSize.
.PP 
.TP
.B \-c cutoff
Report only base pairs with an average probability > \fIcutoff\fP in the
dot plot; default=0.01.
.PP
.TP
.B \-o
Save memory by printing out everything during computation. Saves memory,
and is activated per default for sequences over 1M bp.
.PP 
.TP
.B \-u (width)
Compute the mean probability that regions of length 1 to \fIwidth\fP are
unpaired. Output is saved in a _lunp file. Default width is 31.
.PP 
.TP
.B \-O
toggles \fB\-u\fP option, output is switched from probabilities to their
logarithm, which are NOT exactly the mean energies needed to the respective
stretch of bases!
.PP 
The \fB\-T\fP, \fB\-d\fP \fB\-4\fP, \fB\-noGU\fP, \fB\-noCloseGU\fP,
\fB\-e\fP, \fB\-P\fP, \fB\-nsp\fP, options work as in
RNAfold. 

.SH "SEE ALSO"
RNALfold(1)

.SH "REFERENCES"
S. H. Bernhart, I.L. Hofacker, and P.F. Stadler (2006)
"Local Base Pairing Probabilities in Large RNAs" 
Bioinformatics 22: 614\-615
.br
A.F. Bompfünewerer, R. Backofen, S.H. Bernhart, J. Hertel, I.L. Hofacker,
P.F. Stadler, S. Will (2007) 
"Variations on {RNA} Folding and Alignment: Lessons from Benasque"
J. Math. Biol. 
.SH "VERSION"
This man page documents version 1.6 Vienna RNA Package.
.SH "AUTHORS"
Stephan H Bernhart, Ivo L Hofacker, Peter F Stadler. 
.br 
Send comments to <rna@tbi.univie.ac.at>


NEWS
See Changelog for details.

Version 1.8
------------
- new RNAalifold has better treatment of gaps and ribosum based covariance
  scores. Use the -old switch for compatibility with older RNAalifold
  versions.
- RNAplfold -u  now computes all accessibilities up to a maximum length
  (much faster than computing each individually)
- ATTENTION: output formats of RNAplfold -u and or RNAup have been changed  
  Programs parsing RNAplfold and RNAup output will have to be modified.

Version 1.7
------------
- RNAplfold can now be used to compute accessibilities, i.e. the
  probability that a stretch of the RNA remains unpaired (and thus
  available for intermolecular interactions).
- A new version of RNAup predicts RNA-RNA interactions takeing into account
  the competition between inter- and intramolecular structure in both
  molcules
- Circular RNAs can be treated by RNAfold, RNAalifold, RNAsubopt, and RNAcofold
- RNAaliduplex predicts RNA-RNA interactions between two sets of aligned
  sequences (inter-molecular structure only)

Version 1.6
------------
- The RNAforester program for tree-alignments of RNA structures is now
  distributed with the Vienna RNA package, see the RNAforester
  subdirectory for more information. RNAforester was written by Matthias
  Hoechsmann <mhoechsm@techfak.uni-bielefeld.de> 
- The Kinfold program for stochastic simulation of folding trajectories is
  now included in the package, see the Kinfold subdirectory.
- cofolding of two structures now supports suboptimal folding and 
  partition function folding. 
  ATTENTION: Energies of hybrid structures now include the Duplex-initiation
  energy, which was neglected in previous version.
- RNAplfold is a partition function variant of RNALfold. It computes the
  mean probability of a (local) base pair averaged over all sequence
  windows that contain the pair.
- new utilities to color alignments and consensus structures  
- RNAfold -p now computes the centroid structure
- ATTENTION: ensemble diversities in version <1.6.5 are off by a factor 2

Version 1.5pre 
-----------
- ViennaRNA now uses autoconfig generated configure scripts for even better
  portability (should compile on any UNIX, Linux, MacOS X, Windows with
  Cygwin).
- The new RNAalifold program predicts consensus structures for a set of
  aligned sequences.
- Complete suboptimal folding is now integrated in the library.
- Beginning support for co-folding of two strands: energy_of_struct() and
  RNAeval can now compute energies of duplex structures.
- RNAcofold predicts hybrid structures of two RNA strands
- RNAduplex predicts hybrid structures, while allowing only inter-molecular
  base pairs (useful for finding potential binding sites)
- RNALfold predicts locally stable structures in long sequences.
- Major changes to Perl module. See the pod documentation (perldoc RNA).
- RNAsubopt can do stochastic backtracking to produce samples of suboptimal
  structures with Boltzmann statistics.
- New utilities to rotate secondary structure plots and annotate them with
  reliability data.
- Various small bug fixes

Version 1.4
-----------
- New Turner parameters as described in Mathews et.al. JMB v288, 1999.
  Small changes to format of parameter files (old param files won't work!)
- mfe and suboptimal folding will produce only structures without isolated
  pairs if noLonelyPairs=1 (-noLP option), for partition function folding
  pairs that can only occur as isolated pairs are not formed.
- setting dangles=3 (-d3 option) will allow co-axial stacking of adjacent
  helices in mfe folding and energy_of_struct().

Version 1.3.1
-------------
- RNAheat would produce spikes in the specific heat because dangling end
  energies did not go smoothly to 0. 
- PS dot plots now have an option to use a log scale (edit _dp.ps file
  and set logscale to true).

Version 1.3
-----------
- Secondary structure plots now use E. Bruccoleri's naview routines for
  layout by default. New utility RNAplot produces secondary structure plots 
  from structures in bracket notation with several options.
- New -d2 option in RNAfold and RNAeval sets dangles=2, which makes 
  energy_of_struct() and fold() treat dangling ends as in pf_fold().
  -noLP option in RNAfold etc sets noLonelyPairs=1, which avoids most
  structures containing lonely base pairs (helices of length 1).
- new utility functions pack_structure() unpack_structure() make_pair_table()
  and bp_distance().  RNAdistance adds bp_distance() via -DP switch.
- First release of RNAsubopt for complete suboptimal folding.
- fixed bug in asymmetry penalty for interior loops. 
- Default compilation now uses doubles for partition function folding.

Version 1.2.1
-------------
- Fixed bug in version 1.2 of the RNAheat program causing overflow errors
  for most input sequences. 
- The PS_dot_plot() and PS_rna_plot() routines now return an int. The return 
  value is 0 if the file could not be written, 1 otherwise.
- This version contains the alpha version of a perl5 module, which let's you
  access all the capabilities of the Vienna RNA library from perl scripts.

Version 1.2
-----------
- New energy parameters from (Walter et.al 1994). 
- Energy parameters can be read from file.
- RNAeval and energy_of_struct() support logarithmic energy function for 
  multi-loops.
- gmlRNA() produces secondary structure drawing in gml (Graph Meta Language). 
- Many bug fixes.


Progs/RNAplfold.c
/* Last changed Time-stamp: <2008-07-02 17:10:04 berni> */
/*
		  Ineractive Access to folding Routines

		  c Ivo L Hofacker
		  Vienna RNA package
*/

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <ctype.h>
#include <unistd.h>
#include <string.h>
#include "fold.h"
#include "part_func.h"
#include "fold_vars.h"
#include "utils.h"
#include "PS_dot.h"
#include "energy_const.h"
#include "LPfold.h"

extern float Lfold(char *string, char *structure, int winsize);
extern void  read_parameter_file(const char fname[]);

/*@unused@*/
static char rcsid[] = "$Id: RNAplfold.c,v 1.10 2008/07/02 15:34:24 ivo Exp $";

#define PRIVATE static

static char  scale[] = "....,....1....,....2....,....3....,....4"
	"....,....5....,....6....,....7....,....8";

PRIVATE void usage(void);
PRIVATE void putout_pup(double *pup,int length, int winsize, char *name);
PRIVATE void putoutpU_G(double **pU,int length, int ulength, FILE *fp);
int unpaired;
PRIVATE void putoutphakim_u(double **pU,int length, int ulength, FILE *fp);
/*--------------------------------------------------------------------------*/

int main(int argc, char *argv[])
{
  char *string, *line;
  char *structure=NULL, *cstruc=NULL;
  char  fname[80], ffname[100];
  char  *ParamFile=NULL;
  char  *ns_bases=NULL, *c;
  int   i, length, l, sym,r;
  int   istty;
  int noconv=0;
  int winsize=70;
  int pairdist=0;
  float cutoff=0.01;
  int tempwin=0;
  int temppair=0;
  int tempunpaired=0;
  FILE *pUfp=NULL;
  FILE *spup=NULL;
  double **pup=NULL; /*prob of being unpaired, lengthwise*/
  int plexoutput=0;
  plist *dpp=NULL;
  plist *pl;
  int simply_putout=0;
  int openenergies=0;
  string=NULL;
  dangles=2;
  unpaired=0;
  for (i=1; i<argc; i++) {
    if (argv[i][0]=='-')
      switch ( argv[i][1] )
	{
	case 'T':  if (argv[i][2]!='\0') usage();
	  if(i==argc-1) usage();
	  r=sscanf(argv[++i], "%lf", &temperature);
	  if (!r) usage();
	  break;
	case 'n':
	  if ( strcmp(argv[i], "-noGU")==0) noGU=1;
	  if ( strcmp(argv[i], "-noCloseGU")==0) no_closingGU=1;
	  if ( strcmp(argv[i], "-noLP")==0) noLonelyPairs=1;
	  if ( strcmp(argv[i], "-nsp") ==0) {
	    if (i==argc-1) usage();
	    ns_bases = argv[++i];
	  }
	  if ( strcmp(argv[i], "-noconv")==0) noconv=1;
	  break;
	case '4':
	  tetra_loop=0;
	  break;
	case 'e':
	  if(i==argc-1) usage();
	  r=sscanf(argv[++i],"%d", &energy_set);
	  if (!r) usage();
	  break;
	case 'c':
	  if(i==argc-1) usage();
	  r=sscanf(argv[++i],"%f", &cutoff);
	  if (!r) usage();
	  break;
/*	case 'C':
	  fold_constrained=1;
	  break;
	case 'S':
	  if(i==argc-1) usage();
	  r=sscanf(argv[++i],"%lf", &sfact);
	  if (!r) usage();
	  break;
*/
	case 'd': dangles=0;
	  if (argv[i][2]!='\0') {
	    r=sscanf(argv[i]+2, "%d", &dangles);
	    if (r!=1) usage();
	    if (!dangles%2) usage(); /*1,3 not possible*/
	  }
	  break;
	case 'P':
	  if (i==argc-1) usage();
	  ParamFile = argv[++i];
	  break;
	case 'O':
	  openenergies=1;
	   break;
	case 'W':
	  if (i==argc-1) usage();
	  r=sscanf(argv[++i], "%d", &winsize);
	  if (r!=1) usage();
	  break;
	case 'u':
	  /*default = 31*/
	  if (i==argc-1) {
	    unpaired=31;
	    fprintf(stderr,"Computing unpaired probabilities with default parameters: unpaired length=%d\n", unpaired);
	  }
	  else {
	    r=sscanf(argv[i+1], "%d", &unpaired);
	    if (r!=1){
	      unpaired=31;
	      fprintf(stderr,"Computing unpaired probabilities with default parameters: unpaired length=%d\n", unpaired);
	    }
	    else {
	      i++;
	    }
	  }
	  if (unpaired<0) usage();
	
	  break;
	case 'L':
	  if (i==argc-1) usage();
	  r=sscanf(argv[++i], "%d", &pairdist);
	  if (r!=1) usage();
	  break;
	case 'o':
	  simply_putout=1;
	  break;
	case 'p':
	  plexoutput=1;
	  break;
	default: usage();
	}
  }

  if (ParamFile != NULL)
    read_parameter_file(ParamFile);

  if (ns_bases != NULL) {
    nonstandards = space(33);
    c=ns_bases;
    i=sym=0;
    if (*c=='-') {
      sym=1; c++;
    }
    while (*c!='\0') {
      if (*c!=',') {
	nonstandards[i++]=*c++;
	nonstandards[i++]=*c;
	if ((sym)&&(*c!=*(c-1))) {
	  nonstandards[i++]=*c;
	  nonstandards[i++]=*(c-1);
	}
      }
      c++;
    }
  }
  if ((openenergies) &&(unpaired==0)) unpaired=31;
  istty = isatty(fileno(stdout))&&isatty(fileno(stdin));
   if (pairdist==0) pairdist=winsize;
  if (pairdist>winsize) {
    fprintf(stderr, "pairdist (-L %d) should be <= winsize (-W %d);"
	    "Setting pairdist=winsize\n",pairdist, winsize);
    pairdist=winsize;
  }
  do {				/* main loop: continue until end of file */
   if (istty) {
      printf("\nInput string (upper or lower case); @ to quit\n");
      printf("%s\n", scale);
    }
    fname[0]='\0';
    if ((line = get_line(stdin))==NULL) break;

    if (tempwin!=0) {
      winsize=tempwin;
      tempwin=0;
    }
    if (temppair!=0) {
      pairdist=temppair;
      temppair=0;
    }
    if (tempunpaired!=0){
      unpaired=tempunpaired;
      tempunpaired=0;
    }
    /* skip comment lines and get filenames */
    while ((*line=='*')||(*line=='\0')||(*line=='>')) {
      if (*line=='>')
	(void) sscanf(line, ">%42s", fname);
      printf("%s\n", line);
      free(line);
      if ((line = get_line(stdin))==NULL) break;
    }

    if ((line ==NULL) || (strcmp(line, "@") == 0)) break;

    string = (char *) space(strlen(line)+1);
    (void) sscanf(line,"%s",string);
    free(line);
    length = (int) strlen(string);
   if (unpaired) {
      pup=(double **)space((length+1)*sizeof(double *));
      pup[0]=(double *)space(sizeof(double)); /*I only need entry 0*/
      pup[0][0]=unpaired;
    }

    structure = (char *) space((unsigned) length+1);

    for (l = 0; l < length; l++) {
      string[l] = toupper(string[l]);
      if (!noconv && string[l] == 'T') string[l] = 'U';
    }
    if (istty)
      printf("length = %d\n", length);

    if (length>1000000) {

      if (!simply_putout) {
	printf("Switched to simple output mode!!!\n");
      }
      simply_putout=1;
      
      }
    /* initialize_fold(length); */
    update_fold_params();
    if (length<winsize) {
      fprintf(stderr, "WARN: window size %d larger than sequence length %d\n",
	      winsize, length);
      tempwin=winsize;
	winsize=length;
	if (pairdist>winsize) {	  
	  temppair=pairdist;
	  pairdist=winsize;
	}
	if (unpaired>winsize) {
	  tempunpaired=unpaired;
	  unpaired=winsize;
	}
    }
    if (length >= 5) {
      pf_scale = -1;
      pUfp=NULL;
      if (simply_putout) /**/ {
	char oname[50];
	if (unpaired!=0) {
	  char aname[50];
	  if (fname[0]!='\0') {
	    strcpy(aname, fname);
	    strcat(aname, "_lunp");
	  }
	  else strcpy(aname, "plfold_lunp");
	pUfp=fopen(aname,"w");
	}
	if (fname[0]!='\0') {
	  strcpy(oname, fname);
	  strcat(oname, "_basepairs");
	}
	else strcpy(oname, "plfold_basepairs");
	spup=fopen(oname,"w");
      }
     
      pl=pfl_fold(string, winsize, pairdist, cutoff, pup, &dpp, pUfp,spup);
      if (pUfp!=NULL) fclose(pUfp);
      if (spup!=NULL)  fclose(spup);
     if (!simply_putout) {
	if (fname[0]!='\0') {
	  strcpy(ffname, fname);
	  strcat(ffname, "_dp.ps");
	}
	else strcpy(ffname, "plfold_dp.ps");
	PS_dot_plot_turn(string, pl, ffname, pairdist);
	free(pl);
	if (unpaired){
	  char aname[50];
	  if(plexoutput) {
	    FILE *POP;
	    char oname[50];
	    if (fname[0]!='\0') {
	      strcpy(oname, fname);
	      strcat(oname, "_uplex");
	    } 
	    else strcpy(oname, "plfold_uplex");
	    POP=fopen(oname,"w");
 	    putoutphakim_u(pup,length, unpaired, POP);
	    fclose(POP);
	  }
	  if (openenergies) {
	    if (fname[0]!='\0') {
	      strcpy(aname, fname);
	      strcat(aname, "_openen");
	    }
	    else strcpy(aname, "plfold_openen");
	  }
	  else {
	    if (fname[0]!='\0') {
	      strcpy(aname, fname);
	      strcat(aname, "_lunp");
	    }
	    else strcpy(aname, "plfold_lunp");
	  }
	  pUfp=fopen(aname,"w");
	  putoutpU_prob(pup, length, unpaired, pUfp,openenergies);
	  fclose(pUfp);
	   
	}
	
      }
     else {
       free(pl);
       if (unpaired) {
	 free(pup[0]);
	 free(pup);
       }
      
     }
      if (cstruc!=NULL) free(cstruc);
    (void) fflush(stdout);
    }
    free(string);
    free(structure);
  } while (1);
  return 0;
}

PRIVATE void usage(void)
{
  nrerror("usage:\n"
	  "RNAplfold [-L span] [-W winsize] [-u size of unpaired region]\n"
	  "          [-T temp] [-4] [-o] [-d[0|1|2]] [-noGU] [-noCloseGU]\n"
	  "          [-noLP] [-P paramfile] [-nsp pairs] [-noconv] [-O]\n");
}

PRIVATE void putout_pup(double *pup,int length, int winsize, char *name) {
  int i;
  float factor;
  float tfact;
  FILE *FP;
  
  FP=fopen(name,"w");
  fprintf(FP,"&prob of being unpaired between i-%d and i\n",unpaired);
  fflush(NULL);
  for (i=unpaired; i<=length; i++) {
    factor=0.;
    if (i<winsize) {
      factor=1./(i-unpaired+1);
    }
    if (i>length-winsize+unpaired-1) {
      tfact=1./(length-i+1);
      if (tfact>factor) {
	factor=tfact;
      }
     
    }
    else {
      tfact=1./(winsize-unpaired+1);
      if (tfact>factor) {
	factor=tfact;
      }
    }
    fprintf(FP,"%d %.6f\n",i,pup[i]*factor);
  }
  fclose(FP);

}
PRIVATE void putoutpU_G(double **pU,int length, int ulength, FILE *fp) {
  /*put out unpaireds */
  int i,k;
  fprintf(fp,"#unpaired probabilities\n #i\tl=");
  for (i=1; i<=ulength; i++) {
    fprintf(fp,"%d\t", i);
  }
  fprintf(fp,"\n");
  for (k=1; k<=length; k++){
    fprintf(fp,"%d\t",k);
    for (i=1; i<=ulength; i++) {
      if (k+(i-1)<=length) fprintf(fp,"%.7g\t",pU[k+(i-1)][i]);
    }
    fprintf(fp,"\n");
    free(pU[k]);
  }
  free(pU[0]);
  free(pU);
  fflush(fp);
}
PRIVATE void putoutphakim_u(double **pU,int length, int ulength, FILE *fp) {
  /*put out Fopen in dekacalories per mol, and F(cond,open) also ind dekacal*/
  int k;
  
  float RT = (temperature+K0)*GASCONST;
  float p0;
  float pdep;
  int f0;
  int fdep;
  
  fprintf(fp,"#energy necessary to unpair as well as to unpair if i-1 is unpaired also, if i+1 is unpaired also in dekacal/mol\n");
  for (k=1; k<=length; k++){
    fprintf(fp,"%d\t",k);   
    p0=pU[k][1];
    f0=(int) -RT*log(p0)/10;
    fprintf(fp,"%d\t", f0);
    if (k>1) {
      pdep=pU[k][2]/pU[k-1][1];
      fdep=(int) -RT*log(pdep)/10;
      fprintf(fp,"%d\t",fdep);


    }
    else  fprintf(fp,"-0\t");
    if (k<length) {
      pdep=pU[k+1][2]/pU[k+1][1];
      fdep=(int) -RT*log(pdep)/10;
      fprintf(fp,"%d\t",fdep);
    }
    else  fprintf(fp,"-0\t");
    fprintf(fp,"\n");
  }
 
  fflush(fp);
}

